src/Evolution/Systems/Cce/NewmanPenrose.hpp

```diff
@@ -69,6 +69,65 @@ struct VolumeWeyl<Tags::Psi0> {
       const Scalar<SpinWeighted<ComplexDataVector, 0>>& one_minus_y);
 };
 
+/*!
+ * \brief Compute the Weyl scalar \f$\Psi_1\f$ in the volume according to the
+ * standard set of Newman-Penrose vectors.
+ *
+ * \details Our convention is \f$\Psi_1 =
+ * l^\alpha n^\beta l^\mu m^\nu C_{\alpha
+ * \beta \mu \nu}\f$.
+ *
+ * \f{align*}{
+ * \Psi_1 =
+ * &\frac{(1-y)^2}{\sqrt{2048} \sqrt{1 + K} R^2}\Bigg\{
+ * 4 J(\bar{\eth }\beta + \tfrac{1}{2} \bar{Q})
+ * - 4(1 + K) (\eth \beta + \tfrac{1}{2} Q) \\
+ * &+(1-y)\Bigg[
+ * 4 (1 + K)\eth\partial_{y}\beta -4 J\bar{\eth }\partial_{y}\beta
+ * + 4 \left(- J \frac{\bar{\eth }' R}{R} + (1 + K) \frac{\eth ' R}{R}\right)
+ * \partial_{y}\beta \\
+ * &\quad + \frac{1}{K}\Bigg(
+ * J\left\{
+ * -2 \partial_{y}\bar{Q} + \partial_{y}\bar{J} [2 (\eth\beta + \tfrac{1}{2}Q) +
+ * J(\bar{\eth}\beta + \tfrac{1}{2} \bar{Q})]
+ * \right\}\\
+ * &\qquad +(1+K)\Big\{
+ * 2 (\partial_{y}Q + J \partial_{y}\bar{Q}) + (\bar{J} \partial_{y}J - J
+ * \partial_{y}\bar{J}) (\eth\beta + \tfrac{1}{2} Q)
+ * \\
+ * &\qquad\quad - (1+K)
+ * [2 \partial_{y}Q + \partial _{y}J (\bar{\eth }\beta + \tfrac{1}{2} \bar{Q})]
+ * \Big\}\Bigg)\Bigg]\Bigg\}
+ * \f}
+ */
+template <>
+struct VolumeWeyl<Tags::Psi1> {
```

Comment:
This needs to be tested. In `tests/Unit/Evolution/Systems/Cce/Test_NewmanPenrose.cpp` we calculate everything for $\Psi_0$ using a Kerr-Schild solution and check that we get zero. You can probably just add the $\Psi_1$ calculation to that.

Reply:
Oops, that's what I get for only grepping inside of `src/`! Ok, I kind of don't want to duplicate all the setup for the psi0 test (which itself is doing a bunch of duplication of work from Test_GaugeTransformBoundaryData). Please advise if I should just ignore this and just duplicate, or if there's a better approach.

Reply:
Since we're going to need to test the other Weyl scalars, and those will require similar spin weighted bondi variables as $\Psi_1$, I'd say make a new function similar to `compute_psi0_of_bh_on_wt()` in `Test_NewmanPenrose.cpp`, except this is only concerned with volume weyl scalars (doesn't do all the boundary stuff necessary for `compute_psi0_of_bh_on_wt()`). Then test $\Psi_0$ and $\Psi_1$ with this new function for now. You'll have to initialize some of the volume variables before hand. I'd have to take a look at how to do that as I don't know off the top of my head

Reply:
Is it possible to factor out all the common code into another helper in `tests/Unit/Helpers/Evolution/Systems/Cce/BoundaryTestHelpers.hpp`? I'm not yet sure what this would look like.

Reply:
I was suggesting `BoundaryTestHelpers.hpp` because a bunch of @Sizheng-Ma's test is copied from a different module, `tests/Unit/Evolution/Systems/Cce/Test_GaugeTransformBoundaryData.cpp`, and both of these modules use helpers from `BoundaryTestHelpers.hpp`; so I wanted to consolidate the code that's common to both tests. But I defer to your/Sizheng's advice for where the common code should live.